consensus/pacemaker.go

```diff
@@ -103,15 +103,17 @@ func (m *paceMaker) SetConsensusModule(c *ConsensusModule) {
 }
 
 func (p *paceMaker) ValidateMessage(m *typesCons.HotstuffMessage) error {
+	currentHeight := p.consensusMod.Height
```

Comment:
I really like the new simpler implementation, but this function is a good example of trying to figure out when I could/should use the lock or not.

Intuitively, I'd assume that I should acquire a read lock at the beginning of this function before getting `currentHeight := p.consensusMod.Height`, and I don't think it'll hurt, but the question is: should I?

Per this comment: `and a RLock when you want to make sure that no other goroutine is changing the values you are trying to read concurrently`, I think the answer is yes but I think that not acquiring a lock here was an explicit decision.

Reply:
Great question that perhaps is going to make us both, and anyone reading us in the future really, how to go about this.

Basically we want to acquire lock as a way of stopping time and therefore other goroutine for changing state of the things we are trying to access/manipulate.

If we lock in the middle of a non-atomic operation we risk finding ourselves in an inconsistent state aka 🤷‍♂️

For example here, if we lock inside `ValidateMessage` and another goroutine changes the state right after, the actions that follow the message validation could not be coherent anymore with that state we just validated and cause all sorts of errors.

Your question made me think about it even further and I realized that I was locking inside the message handlers but really I had to include also the validation and that change is visible here https://github.com/pokt-network/pocket/pull/198/commits/7bb04c3fae2c6d016c198f826058b51446dd64ca 

So, to answer your question: I believe that locking context should be the smallest possible but not smaller than a single "unit of work".
This PR allowed me to dive deeper into the code and better understand its moving parts where I started naively with the smallest locks possible (getters/setters) but then with tests and questions I started zooming out until I found what I believe is now the right balance.

My confidence however increased by running this whole thing every once in a while :smile: :
```bash
for i in {1..100}; do go test -timeout 2s -count=1 -run ^TestTinyPacemakerTimeouts$ github.com/pokt-network/pocket/consensus/consensus_tests; done > test_runpacemakertimeouts.txt 
for i in {1..100}; do go test -timeout 2s -count=1 -run ^TestHotstuff4Nodes1BlockHappyPath$ github.com/pokt-network/pocket/consensus/consensus_tests; done > testrun_happypath.txt 
for i in {1..100}; do go test -timeout 2s -count=1 -race -run ^TestTinyPacemakerTimeouts$ github.com/pokt-network/pocket/consensus/consensus_tests; done > test_runpacemakertimeouts_race.txt 
for i in {1..100}; do go test -timeout 2s -count=1 -race -run ^TestHotstuff4Nodes1BlockHappyPath$ github.com/pokt-network/pocket/consensus/consensus_tests; done > testrun_happypath_race.txt
```

Reply:
> Your question made me think about it even further

Lol, I think my questions have caused you a lot of thinking this week :)

> My confidence however increased by running this whole thing every once in a while

Put it in a make target! Something like `consensus_concurrenct_tests`

> So, to answer your question: I believe that locking context should be the smallest possible but not smaller than a single "unit of work".

Can you add this comment next to the mutex declaration in `consensus/module.go`

Reply:
Addressed in https://github.com/pokt-network/pocket/pull/198/commits/8a30ef9d2911eba6fbf4a34cfb4f15494121d546 and https://github.com/pokt-network/pocket/pull/198/commits/2a7283ecc375cb9520cc1ce8590268fa14c9dffd